app/views/govuk_publishing_components/components/_google_tag_manager_script.html.erb

```diff
@@ -8,6 +8,10 @@
   gtm_attributes = gtm_attributes.join('&')
 %>
 <script>
+dataLayer = [
+  { 'gtm.blacklist' : ['customPixels', 'customScripts', 'html', 'nonGoogleScripts'] },
```

Comment:
Just want to mention that this component is also used by `content-data-admin` and `content-publisher`, so we should check that this change won't cause those applications a problem.

Reply:
Thanks for alerting me - didn't realise. Yes, I'll include devs from those teams to review.

Reply:
Does this work with `blocklist` instead of `blacklist`? My quick search [suggests it should](https://developers.google.com/tag-platform/tag-manager/web/restrict), and it's a more inclusive term to use if possible.

Reply:
Yup, a dataLayer variable is already set in Content Publisher: https://github.com/alphagov/content-publisher/blob/4be9a6b0b2249ce7b513349777d96b684f5e8a73/app/views/layouts/application.html.erb#L7-L14 I presume there is something similar in Content Data.

This would cause a problem as I understand as it will override the previously set dataLayer variable and replace anything already set.

I imagine you've got a choice of a) making this a breaking change and set a new global dataLayer variable or b) checking whether a dataLayer variable exists, and creating it if not before appending to it. (e.g. `window.dataLayer = window.dataLayer || []`)

I don't know a huge amount about how things work but I'm wondering whether b) is the better option as I'm assuming if an app like Content Publisher adds metadata to datalayer after this script has run, it is too late and we've already sent information to tag manager that misses it?

---

minor nit, we're missing a var declaration here

Reply:
> Does this work with `blocklist` instead of `blacklist`? My quick search [suggests it should](https://developers.google.com/tag-platform/tag-manager/web/restrict), and it's a more inclusive term to use if possible.

Yes, quite right @ChrisBAshton - looks like it should. Updated throughout - thanks for spoting.

Reply:
Thanks Graham.
+1 Kevin - [Content Data Admin defines its own `dataLayer` variable too](https://github.com/alphagov/content-data-admin/blob/c1444b5d268a11079e80455c75bef0236308321f/app/views/layouts/application.html.erb#L7-L11). I assume it would be overwritten by the changes in this PR.

Reply:
I think `window.dataLayer` and `var dataLayer` will produce two different things, so you might want to check on that. I know that if `window.dataLayer` exists and `var dataLayer` doesn't, calling `dataLayer` will produce `window.dataLayer`, but not the other way around.

I suspect the best way to do this PR is to add an option to the component to pass an optional blocklist value. The code might look something like this:

```erb
<%
  gtm_auth ||= nil
  gtm_preview ||= nil
  blocklist ||= nil # an array of the form ['customPixels', 'customScripts', 'html', 'nonGoogleScripts']

  gtm_attributes = %w(gtm_cookies_win=x)
    gtm_attributes << "gtm_auth=" + gtm_auth if gtm_auth
    gtm_attributes << "gtm_preview=" + gtm_preview if gtm_preview
  gtm_attributes = gtm_attributes.join('&')
%>
<script>
<% if blocklist %>
  window.dataLayer = window.dataLayer || []
  window.dataLayer.push({ 'gtm.blocklist' : <%= blocklist %> })
<% end %>
var doNotTrack = ( navigator.doNotTrack === '1' || navigator.doNotTrack === 'yes' || navigator.msDoNotTrack === '1' || window.doNotTrack === '1' )
if (!doNotTrack) {
  (function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
  new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
  j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
  'https://www.googletagmanager.com/gtm.js?id='+i+dl+'&<%= raw gtm_attributes %>';f.parentNode.insertBefore(j,f);
})(window,document,'script','dataLayer','<%= gtm_id %>');
}
</script>
```

That way the apps currently using this will stick with the default configuration and not have your new blocklist included, and when you call it in `static` you can pass a blocklist to it.

You'll need to update the component `YML` file to demonstrate the example, and include a test of the erb template. I can help you with those if you want.

Reply:
I like this approach 👍 Thanks @andysellick - will update the code as suggested.

Reply:
PS Still unsure of the inner workings of `dataLayer`, but my [brief experiment](https://github.com/alphagov/govuk_publishing_components/pull/2922#discussion_r960337231) seems to indicate they are either the same or at least synonymous.

Reply:
Thanks @kevindew - great explanation